File: Leetcode Submissions/Problems/Medium/C++/2.cpp
// https://leetcode.com/problems/add-two-numbers/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *head = new ListNode;
        ListNode *current = head;
        int carry = 0, sum;
        while(l1 || l2){
            sum = carry;
            if(l1){
                sum += l1 -> val;
                l1 = l1 -> next;
            }
            if(l2){
                sum += l2 -> val;
                l2 = l2 -> next;
            }
            current -> val = sum % 10;
            carry = sum / 10;
            if(l1 || l2 || carry){
                current -> next = new ListNode;
                current = current -> next;
            }
        }
        if(carry)
            current -> val = carry;
        return head;
    }
};

File: Leetcode Submissions/Problems/Medium/Python/2.py
# https://leetcode.com/problems/add-two-numbers/
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode()
        current = head
        carry = 0
        while l1 is not None or l2 is not None:
            current_sum = carry
            if l1:
                current_sum += l1.val
                l1 = l1.next
            if l2:
                current_sum += l2.val
                l2 = l2.next
            carry = current_sum // 10
            current.val = current_sum % 10
            if carry != 0 or l1 or l2:
                current.next = ListNode()
                current = current.next
        if carry != 0:
            current.val = carry
        return head
